Cubical/HITs/SetTruncation/Properties.agda

```diff
@@ -75,6 +76,121 @@ elim3 Bset g = elim2 (λ _ _ → isSetΠ (λ _ → Bset _ _ _))
                      (λ a b → elim (λ _ → Bset _ _ _) (g a b))
 
 
+-- the recursor for maps into groupoids following the "HIT proof" in:
+-- https://arxiv.org/abs/1507.01150
+module rec→Gpd {A : Type ℓ} {B : Type ℓ'} (Bgpd : isGroupoid B) (f : A → B)
+               (congFConst : ∀ (a b : A) (p q : a ≡ b) → cong f p ≡ cong f q) where
+
+ data H : Type ℓ where
+  η : A → H
+  ε : ∀ (a b : A) → ∥ a ≡ b ∥ → η a ≡ η b -- prop. trunc. of a≡b
+  δ : ∀ (a b : A) (p : a ≡ b) → ε a b ∣ p ∣ ≡ cong η p
+  gtrunc : isGroupoid H
+
+ -- write elimination principle for H
+ module Helim {P : H → Type ℓ''} (Pgpd : ∀ h → isGroupoid (P h))
+              (η* : (a : A) → P (η a))
+              (ε* : ∀ (a b : A) (∣p∣ : ∥ a ≡ b ∥)
+                  → PathP (λ i → P (ε a b ∣p∣ i)) (η* a) (η* b))
+              (δ* : ∀ (a b : A) (p : a ≡ b)
+                  → PathP (λ i → PathP (λ j → P (δ a b p i j)) (η* a) (η* b))
+                          (ε* a b ∣ p ∣) (cong η* p)) where
+
+  fun : (h : H) → P h
+  fun (η a) = η* a
+  fun (ε a b ∣p∣ i) = ε* a b ∣p∣ i
+  fun (δ a b p i j) = δ* a b p i j
+  fun (gtrunc x y p q α β i j k) = isOfHLevel→isOfHLevelDep 3 Pgpd
+                                   (fun x) (fun y)
+                                   (cong fun p) (cong fun q)
+                                   (cong (cong fun) α) (cong (cong fun) β)
+                                   (gtrunc x y p q α β) i j k
+
+ module Hrec {C : Type ℓ''} (Cgpd : isGroupoid C)
+             (η* : A → C)
+             (ε* : ∀ (a b : A) → ∥ a ≡ b ∥ → η* a ≡ η* b)
+             (δ* : ∀ (a b : A) (p : a ≡ b) → ε* a b ∣ p ∣ ≡ cong η* p) where
+
+  fun : H → C
+  fun = Helim.fun (λ _ → Cgpd) η* ε* δ*
```

Comment:
Isn't it often better to define the recursor directly by recursion instead of using the eliminator? I think this gives better computational behavior?